File: apps/android/app/src/main/java/com/sanvya/app/ui/shell/NavPrefs.kt
package com.sanvya.app.ui.shell

import android.content.Context
import android.content.SharedPreferences
import com.sanvya.app.theme.SanvyaIcons
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import org.json.JSONArray
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import android.content.res.Resources
import com.sanvya.app.i18n.S

/**
 * Which destinations sit in the bottom bar's four customizable slots.
 *
 * A faithful port of `apps/web/src/navPrefs.ts`, down to the storage key
 * (`pc_bottomNav`) and the JSON-array shape, so the two never disagree about
 * what a saved preference means.
 */
data class NavCatalogItem(
    val id: String,
    val route: String,
    val glyph: String,
    /**
     * How this item names itself.
     *
     * A function, not a string and not a string key. It used to be both: a
     * `tkey` of `"nav.transactions"` that nothing ever resolved, sitting beside
     * an English `label` that got rendered directly — so the app shipped
     * English to hi and nl while carrying the key that would have fixed it.
     *
     * Holding the typed accessor instead means a renamed key fails to compile
     * rather than falling back to English at runtime, and there is no second
     * string to keep in sync with the first.
     */
    val label: (Resources) -> String,
)

object NavPrefs : KoinComponent {

    /**
     * The 14 destinations eligible for a slot. Home and More are fixed and are
     * deliberately NOT in here, exactly as on web.
     */
    val CATALOG: List<NavCatalogItem> = listOf(
        NavCatalogItem("transactions", "transactions", SanvyaIcons.swapHoriz, S.Translation::navTransactions),
        NavCatalogItem("friends", "splits", SanvyaIcons.groups, S.Translation::navFriends),
        NavCatalogItem("insights", "insights", SanvyaIcons.insights, S.Translation::navInsights),
        NavCatalogItem("accounts", "accounts", SanvyaIcons.accountBalance, S.Translation::navAccounts),
        NavCatalogItem("budgets", "budgets", SanvyaIcons.donutSmall, S.Translation::navBudgets),
        NavCatalogItem("goals", "goals", SanvyaIcons.flag, S.Translation::navGoals),
        NavCatalogItem("recurring", "recurring", SanvyaIcons.autorenew, S.Translation::navRecurring),
        NavCatalogItem("loans", "loans", SanvyaIcons.requestQuote, S.Translation::navLoans),
        NavCatalogItem("investments", "investments", SanvyaIcons.trendingUp, S.Translation::navInvestments),
        NavCatalogItem("cards", "cards", SanvyaIcons.creditCard, S.Translation::navCards),
        NavCatalogItem("statements", "statements", SanvyaIcons.description, S.Translation::navStatements),
        NavCatalogItem("search", "search", SanvyaIcons.search, S.Translation::navSearch),
        NavCatalogItem("assistant", "assistant", SanvyaIcons.autoAwesome, S.Translation::navAssistant),
        NavCatalogItem("settings", "settings", SanvyaIcons.settings, S.Translation::navSettings),
    )

    val DEFAULT_IDS = listOf("transactions", "accounts", "friends", "insights")
    const val SLOTS = 4

    private const val PREFS_NAME = "sanvya_prefs"
    private const val KEY = "pc_bottomNav"

    private val context: Context by inject()
    private val sharedPrefs: SharedPreferences by lazy {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    private val _ids: MutableStateFlow<List<String>> by lazy { MutableStateFlow(read()) }
    val ids: StateFlow<List<String>> get() = _ids

    fun setIds(ids: List<String>) {
        val clean = sanitize(ids)
        sharedPrefs.edit().putString(KEY, JSONArray(clean).toString()).apply()
        _ids.value = clean
    }

    fun itemsFor(ids: List<String>): List<NavCatalogItem> =
        ids.mapNotNull { id -> CATALOG.firstOrNull { it.id == id } }

    private fun read(): List<String> {
        val raw = sharedPrefs.getString(KEY, null) ?: return DEFAULT_IDS
        return try {
            val array = JSONArray(raw)
            sanitize((0 until array.length()).mapNotNull { array.optString(it, null) })
        } catch (_: Exception) {
            DEFAULT_IDS
        }
    }

    /**
     * Port of web's `sanitize()` — drop unknown ids, dedupe, take the first four,
     * fall back to the defaults when nothing survives, and **top up from the
     * defaults when the list is short**.
     *
     * That last step is not defensive padding. A preference saved before the bar
     * grew to four slots holds three ids, and rendering it as-is leaves one side
     * of the bar short — the lopsided layout this design replaced. Keep it.
     */
    fun sanitize(ids: List<String>): List<String> {
        val valid = ids.filter { id -> CATALOG.any { it.id == id } }
        val deduped = valid.distinct().take(SLOTS).toMutableList()
        if (deduped.isEmpty()) return DEFAULT_IDS
        for (id in DEFAULT_IDS) {
            if (deduped.size >= SLOTS) break
            if (id !in deduped) deduped.add(id)
        }
        return deduped
    }
}

File: apps/ios/App/Shell/NavPrefs.swift
import Foundation
import SwiftUI

/**
 Which destinations sit in the bottom bar's four customizable slots.

 A faithful port of `apps/web/src/navPrefs.ts`, down to the storage key
 (`pc_bottomNav`) and the JSON-array shape, so web, Android and iOS never
 disagree about what a saved preference means.
 */
/// `Equatable` is gone with the string label — a closure cannot be compared,
/// and nothing needed it: the catalog is a constant and rows are keyed by `id`.
struct NavCatalogItem: Identifiable, Sendable {
    let id: String
    let tab: NavTab
    let glyph: String
    /**
     How this item names itself.

     A closure, not a string and not a string key. It used to be both: a `tkey`
     of `"nav.transactions"` that nothing ever resolved, sitting beside an
     English `label` that got rendered directly — so the app shipped English to
     hi and nl while carrying the key that would have fixed it.

     Holding the typed accessor instead means a renamed key fails to compile
     rather than falling back to English at runtime, and there is no second
     string to keep in sync with the first.
     */
    let label: @Sendable () -> String
}

@MainActor
final class NavPrefs: ObservableObject {
    static let shared = NavPrefs()

    /// The 14 destinations eligible for a slot. Home and More are fixed and
    /// deliberately absent, exactly as on web.
    static let catalog: [NavCatalogItem] = [
        NavCatalogItem(id: "transactions", tab: .transactions, glyph: SanvyaIcons.swapHoriz, label: { S.Translation.navTransactions }),
        NavCatalogItem(id: "friends", tab: .splits, glyph: SanvyaIcons.groups, label: { S.Translation.navFriends }),
        NavCatalogItem(id: "insights", tab: .insights, glyph: SanvyaIcons.insights, label: { S.Translation.navInsights }),
        NavCatalogItem(id: "accounts", tab: .accounts, glyph: SanvyaIcons.accountBalance, label: { S.Translation.navAccounts }),
        NavCatalogItem(id: "budgets", tab: .budgets, glyph: SanvyaIcons.donutSmall, label: { S.Translation.navBudgets }),
        NavCatalogItem(id: "goals", tab: .goals, glyph: SanvyaIcons.flag, label: { S.Translation.navGoals }),
        NavCatalogItem(id: "recurring", tab: .recurring, glyph: SanvyaIcons.autorenew, label: { S.Translation.navRecurring }),
        NavCatalogItem(id: "loans", tab: .loans, glyph: SanvyaIcons.requestQuote, label: { S.Translation.navLoans }),
        NavCatalogItem(id: "investments", tab: .investments, glyph: SanvyaIcons.trendingUp, label: { S.Translation.navInvestments }),
        NavCatalogItem(id: "cards", tab: .cards, glyph: SanvyaIcons.creditCard, label: { S.Translation.navCards }),
        NavCatalogItem(id: "statements", tab: .statements, glyph: SanvyaIcons.description, label: { S.Translation.navStatements }),
        NavCatalogItem(id: "search", tab: .search, glyph: SanvyaIcons.search, label: { S.Translation.navSearch }),
        NavCatalogItem(id: "assistant", tab: .assistant, glyph: SanvyaIcons.autoAwesome, label: { S.Translation.navAssistant }),
        NavCatalogItem(id: "settings", tab: .settings, glyph: SanvyaIcons.settings, label: { S.Translation.navSettings }),
    ]

    static let defaultIds = ["transactions", "accounts", "friends", "insights"]
    static let slots = 4

    private static let key = "pc_bottomNav"
    private let defaults = UserDefaults.standard

    @Published private(set) var ids: [String]

    private init() {
        ids = NavPrefs.read(from: UserDefaults.standard)
    }

    func setIds(_ ids: [String]) {
        let clean = NavPrefs.sanitize(ids)
        if let data = try? JSONSerialization.data(withJSONObject: clean),
           let json = String(data: data, encoding: .utf8) {
            defaults.set(json, forKey: NavPrefs.key)
        }
        self.ids = clean
    }

    func items(for ids: [String]) -> [NavCatalogItem] {
        ids.compactMap { id in NavPrefs.catalog.first { $0.id == id } }
    }

    private static func read(from defaults: UserDefaults) -> [String] {
        guard let json = defaults.string(forKey: key),
              let data = json.data(using: .utf8),
              let raw = try? JSONSerialization.jsonObject(with: data) as? [Any]
        else { return defaultIds }
        return sanitize(raw.compactMap { $0 as? String })
    }

    /**
     Port of web's `sanitize()` — drop unknown ids, dedupe, take the first four,
     fall back to the defaults when nothing survives, and **top up from the
     defaults when the list is short**.

     That last step is not defensive padding. A preference saved before the bar
     grew to four slots holds three ids, and rendering it as-is leaves one side
     of the bar short — the lopsided layout this design replaced.
     */
    static func sanitize(_ ids: [String]) -> [String] {
        let valid = ids.filter { id in catalog.contains { $0.id == id } }
        var deduped: [String] = []
        for id in valid where !deduped.contains(id) && deduped.count < slots {
            deduped.append(id)
        }
        if deduped.isEmpty { return defaultIds }
        for id in defaultIds where deduped.count < slots && !deduped.contains(id) {
            deduped.append(id)
        }
        return deduped
    }
}
